core/src/main/java/com/linecorp/armeria/common/logging/BinaryContentPreviewer.java

```diff
@@ -111,8 +117,11 @@ public String produce() {
             return produced;
         }
         try {
-            return produced = reproduce(headers, Unpooled.wrappedBuffer(bufferList.toArray(
-                    BYTE_BUFS)).asReadOnly());
+            ByteBuf buffer = Unpooled.wrappedBuffer(bufferList.toArray(BYTE_BUFS));
+            if (buffer.readableBytes() > maxAggregatedLength) {
+                buffer = buffer.slice(0, maxAggregatedLength);
```

Comment:
Just curious. Could we can keep buffers up to the `maxAggregatedLength` bytes? If so, we may make here simple.

Reply:
As @trustin mentioned at [here](https://github.com/line/armeria/pull/1604#discussion_r257870372), I thought it makes sense that a user may be surprised if they got more than expected.

Reply:
I think keeping exactly up to `maxAggregatedLength` bytes may or may not be beneficial depending on the length of the `HttpData` given to `onData()`. How about this:

- Make a copy of `HttpData` if `(maxAggregatedLength - aggregatedLength) * 2 >= data.length()` (or use whatever heuristic you prefer). Otherwise, use `data.slice(data.readerIndex(), maxAggregatedLength - aggregatedLength)`.
- When creating a composite buffer using `Unpooled.wrappedBuffer()`, then we do not need to create a slice.

Reply:
> As @trustin mentioned [here](https://github.com/line/armeria/pull/1604#discussion_r257870372), I thought it makes sense that a user may be surprised if they got more than expected.

Maybe my comment wasn't clear. I meant that the `bufferList` may keep the bytes up to the `maxAggregatedLength`. So a user will not get more bytes than expected. :-)

Reply:
> > As @trustin mentioned [here](https://github.com/line/armeria/pull/1604#discussion_r257870372), I thought it makes sense that a user may be surprised if they got more than expected.
> 
> Maybe my comment wasn't clear. I meant that the `bufferList` may keep the bytes up to the `maxAggregatedLength`. So a user will not get more bytes than expected. :-)

Oops, Sorry! I got it.